File: stacks_and_queues/strings/c++/backspace_compare/backspace_compare.cpp
// backspace_compare.cpp
#include "backspace_compare.hpp"
#include <string>

CompareStrings::CompareStrings() {}

std::string CompareStrings::buildString(const std::string &string)
{
    std::string output;
    output.reserve(string.length());
    for (const auto &character : string)
    {
        // If '#' and the stack is not empty, pop from the top of the stack
        if (!output.empty() && character == '#')
        {
            output.pop_back();
        }
        else if (character != '#')
        {
            output.push_back(character);
        }
    }
    return output;
}

bool CompareStrings::compare(const std::string &s, const std::string &t)
{
    if (s.length() == 1 && t.length() == 1)
    {
        return s == t;
    }
    return this->buildString(s) == this->buildString(t);
}

File: stacks_and_queues/stacks/strings/python3/backspace_compare.py
from typing import List


class CompareStrings(object):
    """
    A class to compare two strings that contain backspaces.
    """

    def _build_string(self, string: str) -> str:
        """
        Build a string from the given string by pushing and popping characters from a stack.

        Parameters
        ----------
        string : str
            The string to build

        Returns
        -------
        str
            The built string
        """
        char_stack: List[str] = []
        for char in string:
            # If '#' and the stack is not empty, pop the top of the stack
            if char == "#" and char_stack:
                char_stack.pop(-1)
            # Only push if char is not '#'
            elif char != "#":
                char_stack.append(char)
        return "".join(char_stack)

    def compare(self, s: str, t: str) -> bool:
        """
        Compare two strings that contain backspaces.

        Parameters
        ----------
        s : str
            The first string
        t : str
            The second string

        Returns
        -------
        bool
            `True` if the strings are the same, `False` otherwise
        """
        if len(s) == 1 and len(t) == 1:
            return s == t
        return self._build_string(s) == self._build_string(t)


def main() -> int:
    test_strings = [("y#fo##f", "y#f#o##f"), ("a#c", "b"), ("ab#c", "ad#c")]
    for s, t in test_strings:
        compare_strings = CompareStrings()
        same = compare_strings.compare(s, t)
        print(
            f"Given s = {s} and t = {t}, the built strings are {'the' if same else 'not the'} same"
        )

    return 0


if __name__ == "__main__":
    main()
